File: 백준/Silver/11048. 이동하기/이동하기.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
maze = [list(map(int, input().split())) for _ in range(n)]
dp = [[0] * m for _ in range(n)]

for i in range(n):
    for j in range(m):
        if i == 0 and j == 0:
            dp[0][0] = maze[0][0]

        else:
            # 맨 위쪽
            if i == 0:
                dp[0][j] = maze[0][j] + dp[0][j-1]
            # 맨 왼쪽
            elif j == 0:
                dp[i][0] = maze[i][0] + dp[i-1][0]

            else:
                dp[i][j] = maze[i][j] + max(dp[i-1][j], dp[i][j-1])

print(dp[n-1][m-1])

File: 백준/Silver/11048. 이동하기/이동하기.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder output = new StringBuilder();
	static StringTokenizer tokens;

	public static void main(String[] args) throws IOException {
		tokens = new StringTokenizer(input.readLine());
		int N = Integer.parseInt(tokens.nextToken());
		int M = Integer.parseInt(tokens.nextToken());
		int[][] map = new int[N+1][M+1];
		int[][] dp = new int[N+1][M+1];
		
		for(int r = 1; r <= N; r++) {
			tokens = new StringTokenizer(input.readLine());
			for(int c = 1; c <= M; c++) {
				map[r][c] = Integer.parseInt(tokens.nextToken());
			}
		}
		
		dp[1][1] = map[1][1];
		
		for(int r = 1; r <= N; r++) {
			for(int c = 1; c <= M; c++) {
				dp[r][c] = map[r][c] + Math.max(Math.max(dp[r-1][c], dp[r][c-1]), dp[r-1][c-1]);
			}
		}
		
		System.out.println(dp[N][M]);
	}

}